File: Python/data_structures/hash_table.py
# JavaScript Way translated to Python

class HashTable:
    def __init__(self, size):
        self.data = [None for elem in range(size)]

    def hash_function(self, key):
        _hash = 0
        for i in range(len(key)):
            _hash = (_hash + ord(key[i]) * i) % len(self.data)
        return _hash

    def set(self, key, value):
        point = self.hash_function(key)
        if not self.data[point]:
            self.data[point] = []
            self.data[point].append([key, value])
        else:
            self.data[point].append([key, value])
        return True

    def get(self, key):
        point_data = self.data[self.hash_function(key)]
        for i in range(len(point_data)):
            if point_data[i][0] == key:
                return point_data[i][1]
        return point_data[1]

    def keys(self):
        keys_array = []
        for i in range(len(self.data)):
            if self.data[i]:
                for j in range(len(self.data[i])):
                    keys_array.append(self.data[i][j][0])
        return keys_array


print("JavaScript Way translated to Python")
myHashTable = HashTable(10)
myHashTable.set('grapes', 1000)
myHashTable.set('lemon', 100)
myHashTable.set('apple', 330)
myHashTable.set('orange', 54)
print(myHashTable.get('apple'))
print(myHashTable.keys())

# Pythonic code


class PyHashTable:
    def __init__(self, size):
        self.data = [None for elem in range(size)]

    def hash_function(self, key):
        _hash = 0
        for i in range(len(key)):
            _hash = (_hash + ord(key[i]) * i) % len(self.data)
        return _hash

    def set(self, key, value):
        point = self.hash_function(key)
        if not self.data[point]:
            self.data[point] = []
            self.data[point].append([key, value])
        else:
            self.data[point].append([key, value])
        return True

    def get(self, key):
        point_data = self.data[self.hash_function(key)]
        for point_key in point_data:
            if point_key == key:
                return point_key[1]
        return point_data[0][1]

    def keys(self):
        keys_array = []
        for point_key in self.data:
            if point_key:
                for value in point_key:
                    keys_array.append(value[0])
        return keys_array


print("")
print("Pytonic way")
myHashTable = PyHashTable(10)
myHashTable.set('grapes', 1000)
myHashTable.set('lemon', 100)
myHashTable.set('apple', 330)
myHashTable.set('orange', 54)
print(myHashTable.get('apple'))
print(myHashTable.keys())

File: javascript/DataStructures/HashTable/hashTable.js
//Hash Tables = Hash Maps, Maps, Unordered Maps, Dictionaries, Objects...
// Have the key -> Value pair.
//Hash function is simply a function tha generates a value of fixed length for each input that it gets. Examples: md5, sha1, sha256. 
//Idempotent: A fancy way of saying that a function, given an input, always outputs the same output, and underneath the hood, because hash tables exists in all languages, they're implemented with an optimum hashing function that's really, really fast, with time complexity of O(1).

//Time Complexity: Insert -> O(1); lookup -> O(1); delete -> O(1); search -> O(1).

// ----- Objects -----
let user = {
   age: 54,
   name: 'kylie',
   magic: true,
   scream: function () {
      console.log('ahhhhhhh!')
   }
}

user.scream();
user.spell = 'abra kadabra'; //O(1)
console.log(user.age, user.spell); //O(1)

//Hash Collisions = https://www.cs.usfca.edu/~galles/visualization/OpenHash.html
// When a collision happens our time complexity is O(n/k) where K is the size for our hash table. Removing constants it's became O(n).

// ----- Map -----
// Using maps you can have any type of object including functions as a key. Another benefit of map is that it maintains insertion order.
const a = new Map();

// ----- Sts -----
//Is another feature very similar to Map, the only difference is that it only stores the keys.
const b = new Set();

class HashTable {
   constructor(size) {
      this.data = new Array(size);
   }

   _hash(key) {
      let hash = 0;
      for (let i = 0; i < key.length; i++) {
         hash = (hash + key.charCodeAt(i) * i) % this.data.length;
      }
      return hash;
   }

   set(key, value) {
      let point = this._hash(key);
      if (!this.data[point]) {
         this.data[point] = [];
         this.data[point].push([key, value]);
      }
      else
         this.data[point].push([key, value]); //Collision
      console.log(this.data);
      return true;
   }

   get(key) {
      const pointData = this.data[this._hash(key)];
      for (let i = 0; i < pointData.length; i++) { //Collision
         if (pointData[i][0] === key)
            return pointData[i][1];
      } //O(1) for most of cases but O(n) in case of collision. 
      return undefined;
   }

   keys() {
      const keysArray = []
      for (let i = 0; i < this.data.length; i++) {
         if (this.data[i]) {
            for (let j = 0; j < this.data[i].length; j++)
               keysArray.push(this.data[i][j][0]);
         }
      } //O(n) with no order.
      return keysArray;
   }

}

const myHashTable = new HashTable(10);
myHashTable.set('grapes', 1000);
myHashTable.set('chuchu', 100);
myHashTable.set('apple', 330);
myHashTable.set('orange', 54);
console.log(myHashTable.get('apple'));
console.log(myHashTable.keys());
